### PR TITLE
Fix bug with multilingual engines causing different translations

### LINKED ISSUE #527: Rare multibyte characters sometimes cause irrecoverable error
Hi, 

When submitting a query that contains rare multibyte characters, we get a HTTP 500 error for an **ArrayIndexOutOfBoundsException** , as you can see from the examples below.  This is not always the case, but it is definitely annoying.  Note that this also does not happen with every trained model.

Is this a bug?  Do you have any suggestions about how we could prevent this?

Thank you for looking into the matter.

**Greek capitalize O micron**
localhost:8045/translate?source=en&target=pb&q=Ο
{"status":500,"error":{"type":"**ArrayIndexOutOfBoundsException**"}}

**Unicode range for Enclosed Alphanumerics   [③, ④ -  ⑬, ⑯]**

- Some of them work fine (1, 2, and 3 -  which are also present in the training data)

localhost:8045/translate?source=en&target=pb&q=③
=> {"status":200,"data":{"translation":"③","sourceWordCount":1,"targetWordCount":1,"priority":"normal","totalTime":38,"memoryLookupTime":0,"decodingTime":38,"queueTime":0,"queueLength":0}}

- but 4-12 fail miserably

localhost:8045/translate?source=en&target=pb&q=④
=> {"status":500,"error":{"type":"**ArrayIndexOutOfBoundsException**"}}

- then 13-15 do not give an error, but result in a number from 9324 to 9326

localhost:8045/translate?source=en&target=pb&q=⑬
=> {"status":200,"data":{"translation":"9324;","sourceWordCount":1,"targetWordCount":1,"priority":"normal","totalTime":55,"memoryLookupTime":0,"decodingTime":55,"queueTime":0,"queueLength":0}}

- and from 16 on, the errors start coming again

localhost:8045/translate?source=en&target=pb&q=⑯
=> {"status":500,"error":{"type":"**ArrayIndexOutOfBoundsException**"}}

FYI: the only line shown in the node.log looks like:
2020-01-16 15:03:56,400 ERROR eu.modernmt.api.actions.translation.Translate [qtp1579139754-402] Internal error while executing action Translate
java.lang.ArrayIndexOutOfBoundsException

### LINKED ISSUE #531: Different translation for language pair when used in multilingual engine
I have run into an issue, which I originally discovered in ModernMT 4.2, but which also affects ModernMT 4.5, albeit with a slightly different outcome. In ModernMT 4.2, the error came to light in the form of standalone language pair engine being able to translate a segment just fine, but failing with a 500 error codes in a multilingual setting. It mostly occured for segments containing extremely rare characters, similar to the description in #527 , but I now also experienced an example for FR -> EN with a somewhat standard segment: 
```
Le succès de l'assainissement de l'entité juridique suisse d'une banque d'importance systémique serait ainsi plus probable, ce qui contribuerait à atteindre l'objectif principal, à savoir le maintien ininterrompu des fonctions systémiques.
```
When running the FR -> EN by itself, the result is nice with both MMT 4.2 and 4.5:
```
The success of the reorganisation of the Swiss legal entity of a systemically important bank would thus be more likely, which would contribute to achieving the main objective of uninterrupted maintenance of systemic functions.
```
When using a multilingual engine with MMT 4.2 and a couple of additional language pairs, this results in a 500 error. For some segments, we receive no stacktrace in the logs, for other we get the following:
```
java.lang.ArrayIndexOutOfBoundsException: 33
	at eu.modernmt.processing.WordTextGuessingProcessor.call(WordTextGuessingProcessor.java:21)
	at eu.modernmt.processing.WordTextGuessingProcessor.call(WordTextGuessingProcessor.java:11)
	at eu.modernmt.processing.ProcessingPipeline.call(ProcessingPipeline.java:26)
	at eu.modernmt.processing.VoidProcessingPipeline.call(VoidProcessingPipeline.java:17)
	at eu.modernmt.processing.VoidProcessingPipeline.call(VoidProcessingPipeline.java:9)
	at eu.modernmt.processing.concurrent.PipelineExecutor.process(PipelineExecutor.java:35)
	at eu.modernmt.processing.concurrent.PipelineExecutor.process(PipelineExecutor.java:28)
	at eu.modernmt.processing.Postprocessor.process(Postprocessor.java:45)
	at eu.modernmt.facade.TranslationFacade.get(TranslationFacade.java:110)
	at eu.modernmt.facade.TranslationFacade.get(TranslationFacade.java:73)
	at eu.modernmt.api.actions.translation.Translate.execute(Translate.java:43)
	at eu.modernmt.api.actions.translation.Translate.execute(Translate.java:25)
	at eu.modernmt.api.framework.actions.ObjectAction.getResult(ObjectAction.java:13)
	at eu.modernmt.api.framework.actions.ObjectAction.getResult(ObjectAction.java:8)
	at eu.modernmt.api.framework.actions.JSONAction.unsecureExecute(JSONAction.java:52)
	at eu.modernmt.api.framework.actions.JSONAction.execute(JSONAction.java:23)
	at eu.modernmt.api.framework.routing.RouterServlet.service(RouterServlet.java:85)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:873)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:542)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:205)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:480)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:144)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at eu.modernmt.api.MultipartConfigInjectionHandler.handle(MultipartConfigInjectionHandler.java:49)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at org.eclipse.jetty.server.Server.handle(Server.java:505)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:370)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:267)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:305)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:698)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:804)
	at java.lang.Thread.run(Thread.java:748)
```

When using a multilingual engine in ModernMT 4.5 instead, it "works" in the sense of not crashing, but the translation is the following:
```
&lt;UNK&gt; &lt;UNK&gt; &lt;UNK&gt; &lt;UNK&gt; &lt;UNK&gt; &lt;UNK&gt; de financiële zekerheid van de systemically important bank would be more likely, which would contribute to achieving the main objective of uninterrupted maintenance of systemic functions.
```
I imagine there is no "easy fix" without a new release of Modern MT, but is this something you have experienced before and are possibly already aware of?

### PR DESCRIPTION
**Pull request by ecly**
This bugfix relates to and closes #531, closes #527.
Both bugs were originally thought to relate to handling or rare words, which was partially true.

The problem arises with how ModernMT handles fitting multiple models of varying vocabulary sizes into the same instance of `TransformerModel` from `fairseq`. MMT currently handles this by fitting all models' weights to the size of the largest embedding matrix, padding with zeros as necessary. This works fine when embedding the input, as the padded tokens never occur in the input. As the embedding matrix is however reused as the parameter for the `output_layer` in FAIRSeq, this now creates several logits with probability 0 for models that aren't the one with the largest vocabulary in the multilingual engine. For the most part, this has limited impact in real life scenarios, outside if small but non-critical changes to the softmax distribution. It can however happen that some model will predict a full distribution of logits with negative values, which all of a sudden makes one of the padded tokens the most likely logit with its value of 0. This causes the errors we see in issues #531 and #527, and used to cause exceptions in MMT 4.2.

To solve this, I suggest one workaround in this PR, although it may be possible to solve this in a more elegant way. The solution is quite simple, by simply overriding the `get_normalized_probs` function in the `TransformerModel` from `fairseq` with a wrapped function that truncates the logits to the original vocabulary size of the model, before calling the original `get_normalized_probs` function. It then subsequently pads the probability distribution returned by `get_normalized_probs` with `-math.inf` for all the padded tokens (in a similar way to how the <PAD> token is handled in fairseq's `SequenceGenerator`. This should result in inference being treated the exact same way for engines consisting of one/many models.

I have tested this fix and found that it works for both regular translation and for translation with context memories, without any noticable impacts on throughput.

### CODE REVIEW

**Comment by davidecaroselli** [comment]
Hi @ecly this is really a great PR!
Thank you for having spotted this bug, we will make tests on our side and merge it asap.
